tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,366 @@
+
+from tensorflow.python.ops import summary_ops_v2
+from tensorflow.python.summary import summary as summary_ops_v1
+from tensorflow_model_optimization.python.core.sparsity.keras import pruning_utils
+from tensorflow_model_optimization.python.core.sparsity_tf2 import schedule as update_schedule
+from tensorflow_model_optimization.python.core.sparsity_tf2 import pruner
+from tensorflow_model_optimization.python.core.sparsity_tf2 import sparse_utils
+
+
+class RiGLPruner(pruner.Pruner):
+  """
+  Implementation of the RiGL dynamic sparse training algorithm.
+  """
+
+  def __init__(self,
+      update_schedule=update_schedule.ConstantSchedule(0.1, 0),
+      sparsity=0.5,
+      block_size=(1,1),
+      block_pooling_type='AVG',
+      seed=0,
+      noise_std=0,
+      reinit=False,
+      grow_init='zeros'
+    ):
+    """The logic for magnitude-based RiGL trained weight tensors as presented
+    in https://proceedings.icml.cc/static/paper_files/icml/2020/287-Paper.pdf.
+    At each iteration, according to the update_schedule, connections are grown
+    based on gradient information and dropped based on weight magnitude.
+
+    Args:
+      update_schedule: A `PruningSchedule` object that controls pruning rate
+        throughout training. The first arg is fraction of connections to drop
+        during the update, the second is the starting step to begin updating.
+      sparsity: layer specific sparsity at which the dynamic sparse training method uses
+      block_size: The dimensions (height, weight) for the block sparse pattern
+        in rank-2 weight tensors.
+      block_pooling_type: (optional) The function to use to pool weights in the
+        block. Must be 'AVG' or 'MAX'.
+      seed: assigned by PruningConfig (base added) for multiworker consistency in random processes.
+      reinit: boolean for whether to reinitialize a connection that was dropped and regrown to 
+        its original value, or to set it to 0 otherwise.
+      grow_init: string ('zeros, 'random_normal', 'random_uniform', 'initial_value') way
+        to initialize new connections (i.e. for a connection that was dropped and then grown, 
+        this is to decide how these newly grown weights are initialized),
+    """
+    super(RiGLPruner, self).__init__()
+    self.target_sparsity = sparsity
+    self.update_schedule = update_schedule
+    self._block_size = block_size
+    self.block_pooling_type = block_pooling_type
+    self._seed = seed
+    self._noise_std = noise_std
+    self._drop_regrow_reinit = reinit
+    self._grow_init_method = grow_init
+    if self._grow_init_method not in ('zeros', 'random_normal', 'random_uniform'):
+      raise ValueError('The initialization for growing %s is not a valid option.' % grow_init)
+  
+  def create_slots(self, optimizer, var):
+    base_dtype = var.dtype
+    _deterministic_initializer = sparse_utils.PermuteOnes(self.target_sparsity)
+    kwargs = {'dtype': base_dtype, 'seed': self._seed}
```

Comment:
Doesn't look like you need this indirection, so probably cleaner to inline `dtype=base_dtype` and `seed=self._seed`